File: solution/1457. Pseudo-Palindromic Paths in a Binary Tree/1457.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {
        return dfs(root, 0);
    }
    
private:
    int dfs(TreeNode* node, int ones) {
        if (!node) return 0;
        
        ones ^= 1 << (node->val - 1);
        if (!node->left && !node->right)
            return (ones & (ones - 1)) == 0;
        else
            return dfs(node->left, ones) + dfs(node->right, ones);
    }
};

File: solution/1457. Pseudo-Palindromic Paths in a Binary Tree/1457.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        def dfs(node, x):
            if not node : return 0
            x ^= (1 << node.val)
            if not (node.left or node.right):
                return 1 if x == 0 or (x & (x - 1)) == 0 else 0
            return dfs(node.left, x) + dfs(node.right, x)
            
        return dfs(root, 0)
    
